R/get_ci.R

```diff
@@ -83,95 +167,297 @@ get_ci <- function(dist, interval, estimate, variance = NULL, denominator = NULL
+#' proposed by Tiwari, Clegg, and Zou (2006). Use `"ff97"` for a more
+#' conservative confidence interval proposed by Fay and Feuer (1997).
 #'
-#' @description This function implements the Poisson distribution using the
-#' chi-squared distribution to allow for the continuous extension of integer
-#' variable (`estimate`). Read p.9385 of the [STDRATE Procedure](https://support.sas.com/documentation/onlinedoc/stat/151/stdrate.pdf)
-#' for more information.
+#' @details This function assumes that the values of the input arguments
+#' `estimate`, `weights`, and `variance` are aligned and that they are the same
+#' length.
 #'
-#' @param estimate A scalar, must be a whole number (e.g. number of events)
-#' @param denominator A scalar (e.g. population-time)
-#' @param q_low A scalar, indicating the lower limit
-#' @param q_high A scalar, indicating the upper limit
+#' This function is used to add confidence intervals in [get_ds_rt()].
 #'
-#' @return A numeric vector containing the lower and upper limits
+#' @return A data frame with columns `lower` and `upper` which contain the lower
+#' and upper limits of a confidence interval respectively.
+#'
+#' @references \href{http://documentation.sas.com/doc/en/pgmsascdc/9.4_3.4/statug/statug_stdrate_details.htm}{The STDRATE Procedure}
 #'
 #' @examples
 #' \dontrun{
-#' get_ci_pois(q_low = 0.025, q_high = 0.975, estimate = 30, denominator = 1000)
+#' # calculate a single confidence interval
+#' e_1 <- 0.015
+#' w_1 <- c(
+#'   3.7e-05, 2.2e-05, 2.0e-05, 1.2e-05, 2.3e-05,
+#'   7.0e-05, 3.8e-05, 6.0e-05, 1.7e-05, 2.6e-05
+#' )
+#' v_1 <- 1e-06
+#'
+#' get_ci_gamma(interval = 0.95, estimate = e_1, weights = list(w_1), variance = v_1)
+#' get_ci_gamma(interval = 0.9, estimate = e_1, weights = list(w_1), variance = v_1, method = "ff97")
+#'
+#' # calculate multiple confidence intervals
+#' e_2 <- 0.004
+#' w_2 <- c(
+#'   3.4e-06, 2.0e-06, 1.8e-06, 1.1e-06, 2.1e-06,
+#'   6.3e-06, 3.5e-06, 5.4e-06, 1.5e-06, 2.3e-06
+#' )
+#' v_2 <- 1e-08
+#'
+#' get_ci_gamma(
+#'   interval = 0.9, estimate = c(e_1, e_2),
+#'   weights = list(w_1, w_2), variance = c(v_1, v_2)
+#' )
 #' }
 #'
+#' @importFrom magrittr %>%
+#' @importFrom rlang .data
+#' @export
+get_ci_gamma <- function(interval, estimate, weights, variance, method = "tcz06") {
+  # check validity of inputs
+
+  if (!is.numeric(interval)) {
+    stop("`interval` must be numeric")
+  }
+
+  if (length(interval) != 1) {
+    stop("`interval` must be a scalar")
+  }
+
+  if (interval < 0 | 1 < interval) {
+    stop("`interval` must be between 0 and 1")
+  }
+
+  if (!is.numeric(estimate)) {
+    stop("`estimate` must be numeric")
+  }
+
+  if (!is.list(weights)) {
+    stop("`weights` must be a list")
+  }
+
+  if (!all(sapply(weights, is.numeric))) {
+    stop("`weights` must be a list of numeric vectors")
+  }
+
+  if (!is.numeric(variance)) {
+    stop("`variance` must be numeric")
+  }
+
+  if (!method %in% c("tcz06", "ff97")) {
```

Comment:
Consider relaxing this to take case insensitive inputs

Reply:
Done